### DISCUSSION TITLE
Continuation of a problem, PYModbus can't replicate Modpoll register read

### DISCUSSION THREAD

**Discussion by Hos73** [pymodbus-dev/pymodbus#1538] [Q&A] [answered]
I am continuing a discussion of a problem from stackoverflow, I have a modbus register I can write to with pymodbus and modpoll, however I cannot read from that register with pymodbus, but I can read from it with modpoll.  Discussions that were had on the stackoverflow can be found [here](https://stackoverflow.com/questions/76228494/pymodbus-cant-replicate-modpoll-register-read)
Data sheet of what I am connecting to via usb can be found [here](https://sigrok.org/wiki/ETommens_eTM-xxxxP_Series)
Any assistance is greatly appreciated!
pymodbus version: 3.2.2

**Code is as follows**
```
from pymodbus.client import ModbusSerialClient as RTUClient
import logging
logging.basicConfig()
log = logging.getLogger()
log.setLevel(logging.DEBUG)

PowerSupplyMB = RTUClient(method='rtu', port="/dev/ttyUSB0", baudrate=9600,stopbits=1,bytesize=8,timeout=1)
PowerSupplyMB.connect()
#PowerSupplyMB.write_register(0x0030, int(10.2*100), unit=1)
mb_response  = PowerSupplyMB.read_holding_registers(0x0030, 1, unit=1)

if not mb_response.isError():
    '''isError() method implemented in pymodbus 1.4.0 and above'''
    print(mb_response.registers)  # Your problem is here.

else:
    # Do stuff for error handling.
    print('Error message: {}'.format(mb_response))

```
---
**Output:**
```
DEBUG:pymodbus.logging:Current transaction state - IDLE
DEBUG:pymodbus.logging:Running transaction 1
DEBUG:pymodbus.logging:SEND: 0x0 0x3 0x0 0x30 0x0 0x1 0x85 0xd4
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Transaction failed. (Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received))
DEBUG:pymodbus.logging:Frame - [b''] not ready
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
Error message: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received)
```

**Comment by MattBrittan** [comment]
As suggested on stack overflow; remove the `write_register` call for now. You have established that you can read from the device using  `modpoll` so start by having the python code do exactly the same operation. Once that works you can add extra operations and see if that leads to failure (some devices require a longer delay between requests).

**Reply by Hos73** [reply]
Sorry I should've mentioned and commented that out. But the error that is attached to that is from the version where that write is commented out. 
```
DEBUG:pymodbus.logging:Current transaction state - IDLE
DEBUG:pymodbus.logging:Running transaction 1
DEBUG:pymodbus.logging:SEND: 0x0 0x3 0x0 0x30 0x0 0x1 0x85 0xd4
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Transaction failed. (Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received))
DEBUG:pymodbus.logging:Frame - [b''] not ready
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
Error message: Modbus Error: [Input/Output] Modbus Error: [Invalid Message] No response received, expected at least 4 bytes (0 received)
```

**Reply by MattBrittan** [reply, marked as answer]
OK - Please edit the question to add that. I'd also suggest stating that `modpoll -m rtu -0 -1 -a 1 -b 9600 -p none -r 48 /dev/ttyUSB0` works.

Now looking at your `modpoll` command you include ` -a 1` (Slave address) but looking at the debugging info the request is being sent to slave 0 ("**0x0** 0x3 0x0 0x30 0x0 0x1 0x85 0xd4") so that would explain the lack of response. 

From [the docs](https://pymodbus.readthedocs.io/en/v3.2.2/source/library/client.html#pymodbus.client.mixin.ModbusClientMixin.write_coil) the function definition is [`read_holding_registers(address: int, count: int = 1, slave: int = 0, **kwargs: Any) → ModbusResponse`](https://pymodbus.readthedocs.io/en/latest/source/library/client.html#pymodbus.client.mixin.ModbusClientMixin.read_holding_registers); in your code you say `unit=1` not `slave=1`... (I did not pick this up because you said the `write_register` works - I doubt it actually does).

**Reply by Hos73** [reply]
This did work! The write was properly working as I could visually confirm on the LCD that the voltage was changing. I appreciate the assistance!

**Comment by janiversen** [comment]
Looking at the log, your device does not respond, you are probably using a wrong framer.

**Comment by janiversen** [comment]
rtu framer is the default but many times you need to use the ascii framer.

Apart from that if your device never answers it can be a number of serial parameters that are wrong.

**Reply by janiversen** [reply]
and just for the record I do not follow thread on stackoverflow, normally problems with pymodbus are reported here.

You have a link to the documentation so it should be easy for you to check the serial parameters and the type of framer. I cannot offer that kind of support.

**Reply by janiversen** [reply]
you do not write which version you use, but with never versions your code is wrong, unit= no longer exist, it is slave= (please read our documentation)

**Reply by Hos73** [reply]
Thank you for the response! This did in fact work, I appreciate the assistance!
